### PR TITLE
acceptance test for creating guest user from webui

### PR DESCRIPTION
**Pull request by HariBhandari07**
## Description

Acceptance test for creating guest user form webui
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

**Comment by HariBhandari07** [comment]
needs https://github.com/owncloud/core/pull/35119 to be merged first

File: tests/acceptance/features/bootstrap/WebUIGuestsContext.php [hunk 1]
```diff
@@ -55,21 +58,56 @@ class WebUIGuestsContext extends RawMinkContext implements Context {
...
+	 * @param FilesPage $filesPage
 	 */
-	public function __construct(SetPasswordPage $setPasswordPage) {
+	public function __construct(SetPasswordPage $setPasswordPage, FilesPage $filesPage) {
 		$this->setPasswordPage = $setPasswordPage;
+		$this->filesPage = $filesPage;
+	}
+
+	/**
+	 * @return |null
```
Comment (skshetry) [inline tests/acceptance/features/bootstrap/WebUIGuestsContext.php]:
```suggestion
	 * @return string|null
```

File: tests/acceptance/features/bootstrap/WebUIGuestsContext.php [hunk 2]
```diff
@@ -91,12 +129,57 @@ public function guestUserRegistersUsingWebUI($guestDisplayName, $password) {
...
+	 * @Then user should not be able to share the file with guest user :user
+	 *
+	 * @param string $user
+	 *
+	 * @return void
+	 */
+	public function userShouldNotBeDisplayedInTheDropdownAsGuestUser($user) {
+		$sharingDialogue = $this->filesPage->getSharingDialog();
+		$arrayList = $sharingDialogue->getAutoCompleteItemsList();
+		PHPUnit\Framework\Assert::assertNotContains($this->preUserAddDialog . $user . $this->postUserAddDialog, $arrayList);
```
Comment (skshetry) [inline tests/acceptance/features/bootstrap/WebUIGuestsContext.php]: A suggestion. We used a template in `activity` app to generate strings, like 

https://github.com/owncloud/activity/blob/294b8b745f6ba1178318c1c3eca2477075e27056/tests/acceptance/features/bootstrap/WebUIActivityContext.php#L113

If this is used for creating strings out of template, that might make it more readable.

File: tests/acceptance/features/bootstrap/WebUIGuestsContext.php [hunk 3]
```diff
@@ -91,12 +129,57 @@ public function guestUserRegistersUsingWebUI($guestDisplayName, $password) {
...
+	 * @param string $fileName
+	 * @param string $email
+	 *
+	 * @return void
+	 */
+	public function theUserSharesFileWithGuestUserWithEmailUsingWebui($fileName, $email) {
+		$this->filesPage->openSharingDialog($fileName, $this->getSession());
+		$sharingDialog = $this->filesPage->getSharingDialog();
+		$sharingDialog->shareWithUserOrGroup(
+			$email, $this->preUserAddDialog . $email . $this->postUserAddDialog, $this->getSession()
```
Comment (skshetry) [inline tests/acceptance/features/bootstrap/WebUIGuestsContext.php]: A suggestion. We used a template in `activity` app to generate strings, like 

https://github.com/owncloud/activity/blob/294b8b745f6ba1178318c1c3eca2477075e27056/tests/acceptance/features/bootstrap/WebUIActivityContext.php#L113

If this is used for creating strings out of template, that might make it more readable.

File: tests/acceptance/features/bootstrap/WebUIGuestsContext.php [hunk 4]
```diff
@@ -91,12 +129,57 @@ public function guestUserRegistersUsingWebUI($guestDisplayName, $password) {
...
+		);
+		$this->featureContext->addUserToCreatedUsersList($email, null);
+	}
+
+	/**
+	 * @Then user should not be able to share the file with guest user :user
```
Comment (skshetry) [inline tests/acceptance/features/bootstrap/WebUIGuestsContext.php]:
```suggestion
	 * @Then the user should not be able to share the file with guest user :user
```

File: tests/acceptance/features/bootstrap/WebUIGuestsContext.php [hunk 5]
```diff
@@ -25,6 +25,9 @@
 use Behat\MinkExtension\Context\RawMinkContext;
 use Page\SetPasswordPage;
 use PHPUnit\Framework\Assert;
+use SensioLabs\Behat\PageObjectExtension\PageObject\Exception\ElementNotFoundException;
+use WebDriver\Key;
```
Comment (paurakhsharma) [inline tests/acceptance/features/bootstrap/WebUIGuestsContext.php]: I doesn't seem to be used anywhere. Please remove this.

File: tests/acceptance/features/bootstrap/WebUIGuestsContext.php [hunk 6]
```diff
@@ -91,12 +133,57 @@ public function guestUserRegistersUsingWebUI($guestDisplayName, $password) {
...
+		$this->featureContext->addUserToGroup($user, $this->getGuestGroupName());
+	}
+
 	/**
 	 * @Then a warning should be displayed on the set-password-page saying :expectedMessage
 	 *
 	 * @param string $expectedMessage
 	 *
-	 * @return void
+	 * @return boolean/void
```
Comment (skshetry) [inline tests/acceptance/features/bootstrap/WebUIGuestsContext.php]:
```suggestion
	 * @return boolean|void
```

Comment (skshetry) [reply tests/acceptance/features/bootstrap/WebUIGuestsContext.php]: Is this used anywhere?

File: tests/acceptance/features/bootstrap/WebUIGuestsContext.php [hunk 7]
```diff
@@ -127,5 +214,6 @@ public function before(BeforeScenarioScope $scope) {
 		$this->featureContext = $environment->getContext('FeatureContext');
 		$this->webUIGeneralContext = $environment->getContext('WebUIGeneralContext');
 		$this->guestsContext = $environment->getContext('GuestsContext');
+		$this->sharingContext = $environment->getContext('WebUISharingContext');
```
Comment (paurakhsharma) [inline tests/acceptance/features/bootstrap/WebUIGuestsContext.php]: I doesn't seem to be used anywhere. Please remove this.

File: tests/acceptance/features/bootstrap/WebUIGuestsContext.php [hunk 8]
```diff
@@ -91,12 +133,57 @@ public function guestUserRegistersUsingWebUI($guestDisplayName, $password) {
...
+	 * @param string $fileName
+	 * @param string $email
+	 *
+	 * @return void
+	 */
+	public function theUserSharesFileWithGuestUserWithEmailUsingWebui($fileName, $email) {
+		$this->filesPage->openSharingDialog($fileName, $this->getSession());
+		$sharingDialog = $this->filesPage->getSharingDialog();
+		$sharingDialog->shareWithUserOrGroup(
+			$email, \sprintf($this->userAddDialogBoxFramework, $this->preUserAddDialog . $email . $this->postUserAddDialog), $this->getSession()
```
Comment (skshetry) [inline tests/acceptance/features/bootstrap/WebUIGuestsContext.php]: Save in a separate variable. Too long

File: tests/acceptance/features/bootstrap/WebUIGuestsContext.php [hunk 9]
```diff
@@ -91,12 +133,57 @@ public function guestUserRegistersUsingWebUI($guestDisplayName, $password) {
...
+	 * @Then the user should not be able to share the file with guest user :user
+	 *
+	 * @param string $user
+	 *
+	 * @return void
+	 */
+	public function userShouldNotBeDisplayedInTheDropdownAsGuestUser($user) {
+		$sharingDialogue = $this->filesPage->getSharingDialog();
+		$arrayList = $sharingDialogue->getAutoCompleteItemsList();
+		PHPUnit\Framework\Assert::assertNotContains(\sprintf($this->userRemovedTheShareOfForMsgFramework, $this->preUserAddDialog . $user . $this->postUserAddDialog), $arrayList);
```
Comment (skshetry) [inline tests/acceptance/features/bootstrap/WebUIGuestsContext.php]: Save in a separate variable. Too long

File: tests/acceptance/features/bootstrap/WebUIGuestsContext.php [hunk 10]
```diff
@@ -55,21 +56,62 @@ class WebUIGuestsContext extends RawMinkContext implements Context {
...
+
 	/**
 	 *
 	 * @var SetPasswordPage
 	 */
 	private $setPasswordPage;
 
+	/**
+	 *
+	 * @var PreUserAddDialog
```
Comment (skshetry) [inline tests/acceptance/features/bootstrap/WebUIGuestsContext.php]:
```suggestion
	 * @var string PreUserAddDialog
```

File: tests/acceptance/features/bootstrap/WebUIGuestsContext.php [hunk 11]
```diff
@@ -55,21 +56,62 @@ class WebUIGuestsContext extends RawMinkContext implements Context {
...
+	 */
+	private $preUserAddDialog = 'Add ';
+
+	/**
+	 *
+	 * @var PostUserAddDialog
```
Comment (skshetry) [inline tests/acceptance/features/bootstrap/WebUIGuestsContext.php]:
```suggestion
	 * @var string $postUserAddDialog
```

File: tests/acceptance/features/bootstrap/WebUIGuestsContext.php [hunk 12]
```diff
@@ -55,21 +56,62 @@ class WebUIGuestsContext extends RawMinkContext implements Context {
...
+
+	/**
+	 *
+	 * @var string
+	 */
+	private $userAddDialogBoxFramework = "%s%s%s";
+
+	/**
+	 *
+	 * @var $FilesPage
```
Comment (skshetry) [inline tests/acceptance/features/bootstrap/WebUIGuestsContext.php]:
```suggestion
	 * @var FilesPage $filesPage
```

Comment (HariBhandari07) [reply tests/acceptance/features/bootstrap/WebUIGuestsContext.php]: I followed the style that's already in legacy code:
	 * @var FeatureContext
	 */
	private $featureContext;

	/**
	 *
	 * @var WebUIGeneralContext
	 */
	private $webUIGeneralContext;

	/**
	 *
	 * @var GuestsContext
	 */
	private $guestsContext;

Comment (skshetry) [reply tests/acceptance/features/bootstrap/WebUIGuestsContext.php]: Okay, that works too. But, there's additional `$` there. You can also write following:

```suggestion
	 * @var FilesPage
```

**Comment by individual-it** [comment]
@HariBhandari07 coding standard tests fail. see https://drone.owncloud.com/owncloud/guests/702/42

File: tests/acceptance/features/bootstrap/WebUIGuestsContext.php [hunk 13]
```diff
@@ -91,12 +121,59 @@ public function guestUserRegistersUsingWebUI($guestDisplayName, $password) {
...
+		$this->featureContext->addUserToGroup($user, $this->getGuestGroupName());
+	}
+
 	/**
 	 * @Then a warning should be displayed on the set-password-page saying :expectedMessage
 	 *
 	 * @param string $expectedMessage
 	 *
-	 * @return void
+	 * @return boolean|void
```
Comment (individual-it) [inline tests/acceptance/features/bootstrap/WebUIGuestsContext.php]: I don't think that function ever returns "void" it's ether throws an exception or returns boolean

File: tests/acceptance/features/webUIGuests/guests.feature [hunk 14]
```diff
@@ -27,3 +27,79 @@ Feature: Guests
...
+    Given user "user0" has been created with default attributes
+    And user "user0" has logged in using the webUI
+    When the user shares file "testimage.jpg" with guest user with email "invalid@email.com()9876a" using webUI
+    Then dialog should be displayed on the webUI
+      | title | content               |
+      | Error | Invalid mail address  |
+    And user "invalid@email.com()9876a" should not exist
+
+  @mailhog
+  Scenario: User tries to create a guest user via email with an already used email
```
Comment (individual-it) [inline tests/acceptance/features/webUIGuests/guests.feature]: the test does not use any email address trying to create a guest user

We could use
```
Given these users have been created
  |username | email |
  | user0 |     user0@oc.org |
  | user1 |     user1@oc.org |

.....
When the user types "user1@oc.org" in the share-with-field
```

File: tests/acceptance/features/webUIGuests/guests.feature [hunk 15]
```diff
@@ -27,3 +27,79 @@ Feature: Guests
...
+  @mailhog
+  Scenario: User tries to create a guest user when a server email mode is not set
+    Given user "user1" has been created with default attributes
+    And user "user1" has logged in using the webUI
+    When the administrator deletes system config key "mail_smtpmode" using the occ command
+    And the user shares file "testimage.jpg" with guest user with email "valid@email.com" using webUI
+    Then dialog should be displayed on the webUI
+      | title | content               |
+      | Error | Error while sharing   |
+    And user "invalid@email.com()9876a" should not exist
```
Comment (individual-it) [inline tests/acceptance/features/webUIGuests/guests.feature]: why that email address? It is not used anywhere in the rest of the scenario

File: tests/acceptance/features/bootstrap/WebUIGuestsContext.php [hunk 16]
```diff
@@ -91,12 +121,59 @@ public function guestUserRegistersUsingWebUI($guestDisplayName, $password) {
...
+	 *
+	 * @param string $fileName
+	 * @param string $email
+	 *
+	 * @return void
+	 */
+	public function theUserSharesFileWithGuestUserWithEmailUsingWebui($fileName, $email) {
+		$this->filesPage->openSharingDialog($fileName, $this->getSession());
+		$sharingDialog = $this->filesPage->getSharingDialog();
+		$userAddDialog = sprintf($this->userAddDialogBoxFramework,$email);
```
Comment (skshetry) [inline tests/acceptance/features/bootstrap/WebUIGuestsContext.php]:
```suggestion
		$userAddDialog = sprintf($this->userAddDialogBoxFramework, $email);
```

File: tests/acceptance/features/bootstrap/WebUIGuestsContext.php [hunk 17]
```diff
@@ -91,12 +121,59 @@ public function guestUserRegistersUsingWebUI($guestDisplayName, $password) {
...
+
+	/**
+	 * @Then the user should not be able to share the file with guest user :user
+	 *
+	 * @param string $user
+	 *
+	 * @return void
+	 */
+	public function userShouldNotBeDisplayedInTheDropdownAsGuestUser($user) {
+		$sharingDialogue = $this->filesPage->getSharingDialog();
```
Comment (skshetry) [inline tests/acceptance/features/bootstrap/WebUIGuestsContext.php]: Would be good to be consistent with `Dialogue`/`Dialog`.
```suggestion
		$sharingDialog = $this->filesPage->getSharingDialog();
```

File: tests/acceptance/features/bootstrap/WebUIGuestsContext.php [hunk 18]
```diff
@@ -91,6 +120,53 @@ public function guestUserRegistersUsingWebUI($guestDisplayName, $password) {
...
+		$sharingDialog = $this->filesPage->getSharingDialog();
+		$userAddDialog = \sprintf($this->userAddDialogBoxFramework, $email);
+		$sharingDialog->shareWithUserOrGroup(
+			$email, $userAddDialog, $this->getSession()
+		);
+		$this->featureContext->addUserToCreatedUsersList($email, null);
+	}
+
+	/**
+	 * @Then the user :user should not be displayed in dropdown as guest user
```
Comment (skshetry) [inline tests/acceptance/features/bootstrap/WebUIGuestsContext.php]:
```suggestion
	 * @Then user :user should not be displayed in dropdown as guest user
```

**Comment by skshetry** [comment]
Looks good to me

File: tests/acceptance/features/webUIGuests/guests.feature [hunk 19]
```diff
@@ -7,23 +7,103 @@ Feature: Guests
...
+  @mailhog
+  Scenario: User tries to create a guest user when a server email mode is not set
+    Given user "user1" has been created with default attributes
+    And user "user1" has logged in using the webUI
+    When the administrator deletes system config key "mail_smtpmode" using the occ command
+    And the user shares file "testimage.jpg" with guest user with email "valid@email.com" using webUI
+    Then dialog should be displayed on the webUI
+      | title | content               |
+      | Error | Error while sharing   |
+    And user "invalid@email.com" should not exist
```
Comment (individual-it) [inline tests/acceptance/features/webUIGuests/guests.feature]: why `invalid@email.com`should that not be `valid@email.com`?

File: tests/acceptance/features/webUIGuests/guests.feature [hunk 20]
```diff
@@ -7,23 +7,103 @@ Feature: Guests
...
+    And the user logs in with username "user1" and password "newpassword" using the webUI
+    Then the user should be redirected to a webUI page with the title "Files - %productname%"
+
+  @mailhog
+  Scenario: User tries to create a guest user when a server email is not invalid
+    Given user "user1" has been created with default attributes
+    And user "user1" has logged in using the webUI
+    When the administrator adds system config key "mail_smtphost" with value "conkey" using the occ command
+    And the user shares file "testimage.jpg" with guest user with email "valid@email.com" using webUI
+    Then dialog should be displayed on the webUI
```
Comment (individual-it) [inline tests/acceptance/features/webUIGuests/guests.feature]: check also if user exists

**Pull request merged**